pkg/liqo-controller-manager/resource-request-controller/offer_queue.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2021 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcerequestoperator
+
+import (
+	"context"
+	"crypto/rand"
+	"math/big"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
+)
+
+const (
+	// requeueTimeout define a period of processed items requeue.
+	requeueTimeout = 5 * time.Minute
+
+	// maxRandom is used to generate a random delta to add to requeueTimeout to avoid syncing.
+	maxRandom = 60
+)
+
+// OfferQueue is a component that periodically commands a broadcaster/broker to update its ResourceOffers.
+// It also provides rate-limited retries for transient errors in the offering process.
+type OfferQueue struct {
+	// queue is the underlying generic queue
+	queue        workqueue.RateLimitingInterface
+	// offerUpdater is the OfferUpdater that will create offers
+	offerUpdater *OfferUpdater
+	// identities maps cluster IDs (used by the queue) to ClusterIdentities
+	identities map[string]discoveryv1alpha1.ClusterIdentity
+}
+
+// NewOfferQueue constructs an OfferQueue.
+func NewOfferQueue(offerUpdater *OfferUpdater) OfferQueue {
+	return OfferQueue{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Offer update queue"),
+		offerUpdater: offerUpdater,
+		identities: map[string]discoveryv1alpha1.ClusterIdentity{},
+	}
+}
+
+// Start starts the update loop.
+func (u *OfferQueue) Start(ctx context.Context, group *sync.WaitGroup) {
```

Comment:
Would personally drop the waitgroup and make this block on ctx.Done. Then, it can be added to the manager of controller-runtime.